crates/sui-protocol-constants/src/lib.rs

```diff
@@ -33,14 +33,23 @@ pub const MAX_BASIC_BLOCKS: usize = 1024;
 /// Maximum stack size value. Enforced by the Move bytecode verifier.
 pub const MAX_VALUE_STACK_SIZE: usize = 1024;
 
-/// Maximum number of type nodes. Enforced by the Move bytecode verifier.
+/// Maximum number of "type nodes", a metric for how big a SignatureToken will be when expanded into a fully qualified type. Enforced by the Move bytecode verifier.
 pub const MAX_TYPE_NODES: usize = 256;
 
-/// Maximum number of pushes in one function. Enforced by the Move bytecode verifier.
+/// Maximum number of push instructions in one function. Enforced by the Move bytecode verifier.
 pub const MAX_PUSH_SIZE: usize = 10000;
 
-/// Maximum dependency depth. Enforced by the Move bytecode verifier.
-pub const MAX_DEPENDENCY_DEPTH: u64 = 100;
+/// Maximum number of struct definitions in a module. Enforced by the Move bytecode verifier.
+pub const MAX_STRUCT_DEFINITIONS: usize = 200;
+
+/// Maximum number of function definitions in a module. Enforced by the Move bytecode verifier.
+pub const MAX_FUNCTION_DEFINITIONS: usize = 1000;
+
+/// Maximum number of fields allowed in a struct definition. Enforced by the Move bytecode verifier.
+pub const MAX_FIELDS_IN_STRUCT: usize = 32;
```

Comment:
I copied these limits from https://github.com/move-language/move/pull/723 and they mostly look reasonable to me with the exception of this one (which seems like it might bite a real programmer). The PR summary in 723 says the limit is due to quadratic algos which depend on the # of fields (presumably borrow checker and cyclic struct declaration check?). I'd like to understand this better + make this limit more generous if our bytecode verifier timeout policy can help.

Reply:
I think this exists from the quadratic algorithms in the loader. Namely this one:
https://github.com/move-language/move/blob/main/language/move-vm/runtime/src/loader.rs#L327

More motivation maybe to rethink the loader in the context of Sui?